### PR TITLE
Fix DGC algorithm flow to make it the same as paper

### PR DESCRIPTION
**Pull request by wangxicoding**
At present, the parameter update on DGC is different from the paper. The paper has already made momentum correction in the local and finally uses SGD to update the parameters. While, DGC is using Momentum SGD now, so it needs to fix the algorithm flow of DGC to make it the same as paper.
image

### CODE REVIEW

File: paddle/fluid/operators/optimizers/momentum_op.h [hunk 1]
```diff
@@ -400,5 +400,52 @@ class MomentumOpKernel : public framework::OpKernel<T> {
   }
 };
 
+class MomentumOpMaker : public framework::OpProtoAndCheckerMaker {
```
Comment (gongweibao) [inline paddle/fluid/operators/optimizers/momentum_op.h]: Declare here and need not to define here?

File: python/paddle/fluid/tests/unittests/test_dist_mnist_dgc_nccl.py [hunk 2]
```diff
@@ -33,7 +33,10 @@ def test_dist_train(self):
         if fluid.core.is_compiled_with_cuda():
             self.check_with_place(
                 "dist_mnist.py",
-                delta=1e-5,
+                # FIXME(wangxi): DGC may need a new unit test because its algo flow
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_dist_mnist_dgc_nccl.py]: How to assert when DGC be disabled?

File: paddle/fluid/operators/optimizers/dgc_momentum_op.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      VLOG(10) << " so use momentum optimizer";
+      return _momentum_op_kernel->Compute(context);
+    }
+
+    VLOG(10) << " so use sgd optimizer";
+    return _sgd_op_kernel->Compute(context);
+  }
+
+ private:
+  std::unique_ptr<MomentumOpKernel<DeviceContext, T>> _momentum_op_kernel;
```
Comment (gongweibao) [inline paddle/fluid/operators/optimizers/dgc_momentum_op.h:54]: 👍🏻

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/optimizers/momentum_op.h (hunk 1 above)
Comment (wangxicoding) [reply paddle/fluid/operators/optimizers/momentum_op.h]: Done. Define class declare func in .h，define func in .cc

File: python/paddle/fluid/tests/unittests/test_dist_mnist_dgc_nccl.py (hunk 2 above)
Comment (wangxicoding) [reply python/paddle/fluid/tests/unittests/test_dist_mnist_dgc_nccl.py]: Done. Use log files to judge whether DGC is disabled.

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**